### PR TITLE
add IWithMeta implementation to Reactive* types

### PR DESCRIPTION
**Pull request by ul**
Solution for #28

### CODE REVIEW

**Comment by aaronc** [comment]
So the issue is that once you do this, you've made a copy of your original "ref". I think this would have other undesirable side effects, no? Because of this `clojure.core` opts to have "ref" types use `alter-meta!` instead which does mutate the "ref" but doesn't cause copying the "ref" which could be considered another form of mutation.

**Comment by ul** [comment]
Yes, m.b. it is an issue wanting to use `with-meta` in this situation, but if somebody wants it, it should return copy, because docs say:

> Returns an object of the same type and value as obj, with map m as its metadata.

So we have a choice: if using `with-meta` considered to be harmful, then we should not support using it to not provoke bad code. But is it really harmful? Will initial cursor/atom/expression GCed and returned by `with-meta` one used, or initial c/a/e will coexist with new? Do I correctly understand that first scenario will happen, so using `with-meta` is cheap in this situation?

**Comment by aaronc** [comment]
Well the type of dangerous issue I'm trying to avoid is this:

``` clj
(def a (atom 0))
(dom/mount! root [:p (with-meta a {:binding-disposed (fn [] (do-something))})])
(swap! a inc) ;; oops! the atom that was passed into the view using with-meta isn't the same atom!!!
```

I want metadata on references to be a place where callbacks can be created. But with the `with-meta` solution, you risk creating two different instances. The `alter-meta!` solution also has a potential drawback: you can't create "instances" of the same ref with different metadata - like binding the atom in one place with one callback and in another place with another callback. But since you can't create different "instances" of the same ref anyway, we need another way to work around it.

Basically, `alter-meta!` accomplishes what one would want to do with `with-meta` on reference types because copying a reference type isn't really a meaningful thing to do (i.e. it no longer refers to the same thing so it isn't really a "reference" anymore).... Does that make sense?

**Comment by ul** [comment]
Yes, this kind of dangerous situation seems to be too probable to appear, we should not support `with-meta`
